### PR TITLE
add einsum()

### PR DESCRIPTION
**Pull request by nameless-Chatoyant**
add `einsum()` for some situations.

For example, batch_dot a (?, 1, 3) tensor and a (3, 3) tensor: 
```
K.einsum('ijk,kl->ijl', a, b)
```

which is hard to implement with `batch_dot()`

### CODE REVIEW

**Comment by Dref360** [comment]
Could this be also done for the theano backend?
Could this also support different dim_ordering?
Also add some tests please. If someone change your codw in the future it woukd be nice to check that it didn't break anything.

Great doc byw

File: keras/backend/tensorflow_backend.py [hunk 1]
```diff
@@ -1112,6 +1112,76 @@ def gather(reference, indices):
...
+      (e.g. `ijj,k->ik`).
+    * Subscripts that are summed across multiple inputs (e.g., `ij,ij,jk->ik`).
+
+    # Arguments
+        equation: a `str` describing the contraction, in the same format as
+        `numpy.einsum`.
+        *inputs: the inputs to contract (each one a `Tensor`), whose shapes should
+        be consistent with `equation`.
+
+    # Returns:
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: There should not be a `:` after section titles.

File: keras/backend/tensorflow_backend.py [hunk 2]
```diff
@@ -1112,6 +1112,76 @@ def gather(reference, indices):
...
+        The contracted `Tensor`, with shape determined by `equation`.
+
+    # Raises:
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Same here.

File: keras/backend/tensorflow_backend.py [hunk 3]
```diff
@@ -1112,6 +1112,76 @@ def gather(reference, indices):
...
+        ValueError: If
+            - the format of `equation` is incorrect,
+            - the number of inputs implied by `equation` does not match `len(inputs)`,
+            - an axis appears in the output subscripts but not in any of the inputs,
+            - the number of dimensions of an input differs from the number of
+              indices in its subscript, or
+            - the input shapes are inconsistent along a particular axis.
+
+    # Examples:
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Same here.

**Comment by fchollet** [comment]
Also, we should have a unit test for it in `backend_tests.py`.

**Comment by farizrahman4u** [comment]
Why closed?

**Comment by nameless-Chatoyant** [comment]
`einsum()` can't be done for Theano backend, with Theano's `tensor_dot()` batch_dot performs well. So I think it's a temporary way to make some multiplication as TF remove the `batch_matmul()` but `matmul()` doesn't work well for some situations. I guess it's unnecessary to add `einsum()` to Keras no more.

**Pull request closed without merging**